### ISSUE TITLE
[Bug] 本地重排（rerankModel: local:）完全不可用：tokenizer 传参错误 + worker 结果分发丢失 scores

### ISSUE THREAD

**Issue by minilynn** [Soren-ABT/dsh-knowledge#8]
## 环境

- dsh-knowledge 0.3.4（npm 最新）
- onnxruntime-node 1.21.0 / Node 24.18.0 / Linux x64
- 本地嵌入 + 本地重排：embeddingProvider=local，rerankModel=`local:Xenova/bge-reranker-base`

## 现象：本地重排（`local:` 模式）在 0.3.4 完全不可用

配置 `rerankModel: "local:<模型>"` 后，检索始终 `reranked=false`，且期间经历两种失败形态：

1. **直接崩溃**：`text.replace is not a function`（tokenizer 调用时）
2. **静默失效**：修复第 1 处后不再报错，但重排分数恒为同一值（约 0.0005），检索仍 `reranked=false`，无任何日志

### Bug 1：embed-worker.mjs 中 cross-encoder 的 tokenizer 传参格式错误

`lib/knowledge/embed-worker.mjs` 的 rerank 实现：

```js
// 原代码：把 [query, doc] 作为嵌套数组对传给 tokenizer
const inputs = await tokenizer(batch.map((doc) => [query, doc]), { padding: true, truncation: true });
```

transformers.js 的 tokenizer 不接受这种嵌套对格式（内部对数组调 `.replace` → `text.replace is not a function`）。正确写法是平行数组：

```js
const inputs = await tokenizer(batch.map(() => query), { text_pair: batch, padding: true, truncation: true });
```

> 注：第一次修复只改了 map 部分漏了 `text_pair`，导致输入变成"重复的 query"，分数恒值——两个问题都要一并修。

### Bug 2：index.js 中 worker 结果分发写死了 `message.vectors`

`lib/knowledge/index.js` 主进程收到 worker 消息时的解析：

```js
// 原代码
if (message.ok === true) pending.resolve(message.vectors ?? null);
```

但 worker 对 embed 返回 `{ ok, vectors }`，对 rerank 返回 `{ ok, scores }`——重排时 `message.vectors` 恒为 `undefined`，结果被解成 `null`，下游 `scores2[i]` 抛 TypeError 被搜索路径的 catch 静默吞掉（`reranked` 永远 false，且无日志）。

```js
// 修复：兼容两个字段
if (message.ok === true) pending.resolve(message.vectors ?? message.scores ?? null);
```

## 影响

- `local:` 本地重排（跨编码器）是插件宣城的能力，但 0.3.4 里从未真正工作过
- 用户感知：配置了 rerankModel 后检索结果与未配置完全一样，无任何报错提示

## 建议

1. 修复上述两处（见上）
2. 搜索路径的 rerank catch 建议在 `rescored.length === 0` 时也输出 warn 日志（当前完全静默，排查困难）
3. 可考虑加一个 rerank 自检（检索时若配置了 rerankModel 但从未成功过，提示用户）

## 关联

- 相关 issue：[#5](https://github.com/Soren-ABT/dsh-knowledge/issues/5)（嵌入 worker 空闲回收重拉 self-register）——同一批对本地模型路径的排查中发现

**Comment by Soren-ABT** [comment]
感谢你提供非常详细的定位。经检查确认，`0.3.4` 中确实存在你指出的两处问题，这不是配置错误：

1. cross-encoder tokenizer 错误地接收了嵌套的 `[query, document]` 数组；
2. worker 返回 rerank `scores` 后，主进程仍按 embedding 的 `vectors` 字段解析，导致结果丢失并静默降级。

修复已经合入 `main`：[`[f4dd84c](https://github.com/Soren-ABT/dsh-knowledge/commit/f4dd84c)`](https://github.com/Soren-ABT/dsh-knowledge/commit/f4dd84c)

本次没有只做字段兼容，而是对本地 rerank 链路进行了完整重构：

- tokenizer 改为平行输入：

  ```ts
  tokenizer(batch.map(() => query), {
    text_pair: batch,
    padding: true,
    truncation: true,
  })
  ```

- rerank 从 embedding worker 中分离，使用独立子进程，避免 rerank 卡死或崩溃影响本地 embedding；
- IPC 改为按 `operation`、`id` 和协议版本严格分发，不再使用 `vectors ?? scores` 这种弱类型解析；
- 严格校验 logits/scores 的数量、维度和有限值，异常响应不会被误判为成功；
- 增加超时、队列上限、进程重建、OOM 自动降低 batch 和熔断恢复；
- rerank 失败时保留原始检索结果和顺序，不再静默影响搜索；
- 搜索结果增加结构化 rerank 状态，可区分模型未下载、自检失败、超时、进程崩溃、协议错误等情况；
- 增加模型 readiness、自检和打包校验，确保 `rerank-process.mjs` 被包含在 npm 包中。

适配器、IPC、生命周期、超时恢复、熔断和搜索降级测试已经通过；Linux、Windows、macOS 的无网络协议测试也已覆盖。

包含本次修复的新 npm 版本尚未发布。我们会暂时保持该 Issue 打开，直到完成 Ubuntu Node 22.19/24 的真实 `Xenova/bge-reranker-base` smoke 验证并发布 npm 新版本。

新版本发布后，也希望你能在原来的 Linux 环境中重新验证。如果仍有问题，届时新的 rerank 状态和错误码应该能直接显示具体失败原因，不再出现无日志的静默失效。

再次感谢这份高质量报告。你指出的两个根因都是准确的。

**Comment by Soren-ABT** [comment]
现已完全修复
